File: apps/api/src/main/kotlin/dev/grovarc/api/domain/retrospective/Retrospective.kt
package dev.grovarc.api.domain.retrospective

import dev.grovarc.api.domain.user.User
import jakarta.persistence.*
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.UUID

@Entity
@Table(name = "retrospectives")
class Retrospective(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @Column(nullable = false)
    var title: String,

    @Column(nullable = false, columnDefinition = "TEXT")
    var content: String,

    @Column(nullable = false)
    val periodFrom: LocalDate,

    @Column(nullable = false)
    val periodTo: LocalDate,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: RetrospectiveStatus = RetrospectiveStatus.DRAFT,

    @Column(nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(nullable = false)
    var updatedAt: LocalDateTime = LocalDateTime.now(),
) {
    fun update(title: String, content: String) {
        this.title = title
        this.content = content
        this.updatedAt = LocalDateTime.now()
    }

    fun publish() {
        this.status = RetrospectiveStatus.PUBLISHED
        this.updatedAt = LocalDateTime.now()
    }
}

enum class RetrospectiveStatus {
    DRAFT, PUBLISHED
}

File: apps/web/src/lib/retrospective.ts
import api from "./api";
import type { Retrospective, PageResponse } from "@/types";

export async function getRetrospectives(params?: {
  page?: number;
  size?: number;
  status?: "DRAFT" | "PUBLISHED";
}): Promise<PageResponse<Retrospective>> {
  const { data } = await api.get<PageResponse<Retrospective>>("/api/v1/retrospectives", {
    params,
  });
  return data;
}

export async function getRetrospective(id: string): Promise<Retrospective> {
  const { data } = await api.get<Retrospective>(`/api/v1/retrospectives/${id}`);
  return data;
}

export async function updateRetrospective(
  id: string,
  body: { title?: string; content?: string },
): Promise<Retrospective> {
  const { data } = await api.put<Retrospective>(`/api/v1/retrospectives/${id}`, body);
  return data;
}

export async function publishRetrospective(id: string): Promise<Retrospective> {
  const { data } = await api.post<Retrospective>(`/api/v1/retrospectives/${id}/publish`);
  return data;
}

export async function triggerAiDraft(params: {
  userId: string;
  periodFrom: string;
  periodTo: string;
}): Promise<{ draftTitle: string; draftContent: string; goals: string[] }> {
  const { data } = await api.post("/api/v1/agents/retrospective", {
    user_id: params.userId,
    period_from: params.periodFrom,
    period_to: params.periodTo,
  });
  return data;
}
